rpmio/rpmlog.cc

```diff
@@ -412,3 +415,43 @@ void rpmlog (int code, const char *fmt, ...)
 exit:
     errno = saved_errno;
 }
+
+int rpmlogOnce (uint64_t domain, const char * key, int code, const char *fmt, ...)
+{
+    int saved_errno = errno;
+    rpmlogCtx ctx = rpmlogCtxAcquire();
+    int newkey = 0;
+
+    if (ctx) {
+	wrlock lock(ctx->mutex);
+	newkey = !ctx->seen[domain][{code, key}]++;
```

Comment:
Yeah there's a lot of magic in that line, perhaps a bit too much. This isn't a LoC competition, understandability is 100x more important. If @ffesti himself needed to think for a while to convince himself why this works, that's a danger sign already, maybe this could be opened up a bit, or at least add a comment.

Reply:
Mind you, I don't find this unreadable, just needing a bit of thought too. I guess the thing that raises *my* alarms on first sight is the unchecked [] access which is fine with a map but undefined behavior in some others. Considering we're all relative C++ beginners here, some such remark would not be out of line.

Reply:
Yeah this needs a comment. Is `/* This gets initialized automatically on first access */` clear enough?

Reply:
"This" is unclear :sweat_smile: Maybe something to the tune of "[] access on map creates the element on first access", but the more you think about it the more the comment starts seeming redundant :sweat_smile: 
So dunno - up to you.

Reply:
For me, the most confusing part is the increment vs. negation, it's not clear which one takes precedence (without looking it up in the C++ docs). I assume it's first incremented and only then negated, right?

Reply:
Hmm, actually... it should be the opposite, otherwise it wouldn't work as intended.

Reply:
No, it must be the other way around, hmm. :smile:

Reply:
Just proves Panu's point of this being perhaps a bit confusing :smile:

Reply:
Actually now that I think about it, this follows from the definition of the ++ operator.

Reply:
Maybe what it really needs is a high level comment stating what it does, and leave the understanding as an exercise for the reader as per their C++ experience :sweat_smile: